Automattic/wp-calypso
# Vdimitrakis/try/download stats report

Relates to
- 2644-gh-tumblr/a8c-dsp
- D167420-code

## Proposed Changes
The PR adds a download data button in the campaign item details page (for blaze advertising)

<img width="444" alt="Screenshot 2024-11-21 at 1 04 03 PM" src="https://github.com/user-attachments/assets/f2017e3f-d32c-4e7a-a985-3d5e6f81394e">

When clicked a loading text will appear with text generating report (adding this for translators use)

<img width="402" alt="Screenshot 2024-11-21 at 1 04 06 PM" src="https://github.com/user-attachments/assets/7cb9e79d-43c6-446b-a821-9a07cc869cd3">

internally we call 3 DSP endpoints 

1. Creating a report (`POST /stats/:campaignId/report`) which returns `report_id`;
2. Getting a report status (`GET /stats/:campaignId/report/:reportId/status`), since all the reports are generated using a BullMQ, so we have to know the status of their readiness before being able to download;
3. Getting the actual report (`GET /stats/:campaignId/report/:reportId`), its result has `text/csv` as a Content Type

Nope, we can't just provide a URL of a report to a customer because we use WPCOM proxy, so the request to fetching the report should contain blog- and user-related data.

So, to achieve providing a user with a report we:
- Start report generation by launching its creation,
- Wait until the report is ready by periodically asking for status,
- Fetching the report as text,
- Creating a URL Object to let user download it as a file.

## Testing Instructions
1. Sandbox WPCOM;
2. Apply the diff D167420-code;
3. Run DSP API locally;
4. Run DSP Client locally with `:wpcom: postfix;
5. Adjust Calypso to use the local widget;
6. Make sure you have a campaign with some stats;

If you don't have one, please use this [PR](https://github.tumblr.net/Tumblr/a8c-dsp/pull/2868) from James as an example. In short, you need to:
- have a campaign locally, 
- change its ID to `110111` and start/end dates to be `2024-11-24` and `2024-11-27` respectively,
- run SQL queries from James' PR,
- you can also test that you have data by opening a [link](http://localhost:3003/api/v1.1/stats/110111?start_date=2024-11-24&end_date=2024-11-27&tz=Europe/London&resolution=day) (it should show data for 4 days);

7. Run Calypso locally which is connected to local DSP widget and DSP API is proxied with your WPCOM sandbox;
8. Open the Campaign details in Calypso for "your" campaign `110111`;
9. Click on the "Download" button;
10. Make sure that the title was changed to "Generating a report";
11. Next, your browser should show file downloading;
12. The file is accessible and has correct data.

Demo walk through

https://github.com/user-attachments/assets/1aaaae11-acc1-4fc2-8617-e123a24429c0

## Pre-merge Checklist
